File: 9663_2.c
#include <stdio.h>
#include <stdbool.h>
int N;
int column=0;
int diagonal1=0;
int diagonal2=0;
int result=0;

void place(int placed) {
  if (placed==N) {
    result+=1;
    return;
  }
  for (int i=0; i<N; i++) {
    if (~(column>>i) && ~(diagonal1>>(placed+i)) && ~(diagonal2>>(N-1-placed+i))) {
      column|=1<<i;
      diagonal1|=1<<(placed+i);
      diagonal2|=1<<(N-1-placed+i);
      place(placed+1);
      column&=~(1<<i);
      diagonal1&=~(1<<(placed+i));
      diagonal2&=~(1<<(N-1-placed+i));
    }
  }
}
int main(void) {
  scanf("%d", &N);
  place(0);
  printf("%d", result);
  return 0;
}

File: 9663_2.py
N = int(input())
column = [False]*N
diagonal1 = [False]*(2*N-1)
diagonal2 = [False]*(2*N-1)
result = 0

def place(placed):
  global result
  if placed==N:
    result+=1
    return
  for i in range(N):
    if not column[i] and not diagonal1[placed+i] and not diagonal2[N-1-placed+i]:
      column[i]=diagonal1[placed+i]=diagonal2[N-1-placed+i]=True
      place(placed+1)
      column[i]=diagonal1[placed+i]=diagonal2[N-1-placed+i]=False
place(0)

print(result)
